### PR TITLE
Add some BSP integration tests

### PR DESCRIPTION
**Pull request by alexarchambault**
This adds some BSP integration tests, ensuring the Mill BSP server answers correctly basic BSP requests. BSP responses are serialized on disk and committed, and we compare the requests we get to those on disk. These fixtures on disk can be automatically generated and updated by changing a variable (currently, `BspServerUtil.updateFixtures`), so that it's quite convenient to generate them, and update them and get a nice diff of the changes at the same time.

### CODE REVIEW

**Comment by alexarchambault** [comment]
Not sure the CI error is related to my changes

**Comment by lihaoyi** [comment]
It's the flakiness that should be fixed by https://github.com/com-lihaoyi/mill/pull/3642

**Comment by alexarchambault** [comment]
Don't fear the apparent large diff, it's a lot of test fixtures

File: build.mill
Comment (lefou) [inline build.mill:420]: This is already defined as `TEST_SCALA_2_13_VERSION`.

File: build.mill
Comment (lefou) [inline build.mill:802]: What's the motivation to make this local?

**Review by lefou — COMMENTED** [review]
Great to have some test capabilities. I left some comments below.

File: bsp/worker/src/mill/bsp/worker/MillBuildServer.scala
Comment (lihaoyi) [inline bsp/worker/src/mill/bsp/worker/MillBuildServer.scala:781]: Let's move this into a normal static method rather than an extension method. We don't use extension methods much in com-lihaoyi code unless they're super frequently used (e.g. FastParse's `~`), and `groupList` is used in only two places so definitely doesn't make the cut.

GIven that it's only used in two places, and seems pretty niche and specialized to these two callsites, we can probably also hard-code the `(_._2)(_._1)` thing as well

File: integration/ide/bsp-server/src/BspServerTests.scala
Comment (lihaoyi) [inline integration/ide/bsp-server/src/BspServerTests.scala:40]: This interpolated string is way too long and complicated, let's break it up into some local variables that we combined into a string at the end

File: testkit/src/mill/testkit/BspServerUtil.scala
Comment (lihaoyi) [inline testkit/src/mill/testkit/BspServerUtil.scala:36]: I assume we use Gson here because that's what BSP uses?

File: testkit/src/mill/testkit/BspServerUtil.scala
Comment (lihaoyi) [inline testkit/src/mill/testkit/BspServerUtil.scala:107]: This parameter never seems to be passed; should we just hardcode it as `dummyBuildClient`?

File: testkit/src/mill/testkit/BspServerUtil.scala
Comment (lihaoyi) [inline testkit/src/mill/testkit/BspServerUtil.scala:168]: This name is kind of generic and hard to figure out. How about `normalizePathsForTesting`?

File: testkit/src/mill/testkit/BspServerUtil.scala
Comment (lihaoyi) [inline testkit/src/mill/testkit/BspServerUtil.scala:16]: Let's call this `BspServerTestUtil` so it is clear that this contains test fixtures and not application code helpers

File: testkit/src/mill/testkit/BspServerUtil.scala
Comment (lihaoyi) [inline testkit/src/mill/testkit/BspServerUtil.scala:20]: We should just use a `val` or `def` here unless there is some explicit reason we want it to be lazy (e.g. due to side effecting initialization and/or performance overhead)

File: testkit/src/mill/testkit/BspServerUtil.scala
Comment (lihaoyi) [inline testkit/src/mill/testkit/BspServerUtil.scala:34]: The convention we've used elsewhere in the codebase is `object DummyBuildClient extends DummyBuildClient`

File: testkit/src/mill/testkit/BspServerUtil.scala
Comment (lihaoyi) [inline testkit/src/mill/testkit/BspServerUtil.scala:52]: We can just `os.read` rather than `os.read.bytes`, then we don't need to `new String` later

File: testkit/src/mill/testkit/BspServerUtil.scala
Comment (lihaoyi) [inline testkit/src/mill/testkit/BspServerUtil.scala:50]: I assume `exists -== false` only when generating fixtures for the first time; this should go in a comment or something because it's not obvious

File: integration/ide/bsp-server/src/BspServerTests.scala
Comment (lihaoyi) [inline integration/ide/bsp-server/src/BspServerTests.scala:117]: Let's inline all these `val`s. Right now we have to manually keep track of them and make sure we pass each one to the right `compareWithGsonFixture` call, whereas if inlined then the 1:1 relationship is clear by construction.

**Comment by lihaoyi** [comment]
Left some comments, overall looks reasonable I think

File: testkit/src/mill/testkit/BspServerUtil.scala
Comment (lihaoyi) [inline testkit/src/mill/testkit/BspServerUtil.scala:51]: This name needs to be better. What did we read? Maybe `fixtureValueOpt`?

File: testkit/src/mill/testkit/BspServerUtil.scala
Comment (lihaoyi) [inline testkit/src/mill/testkit/BspServerUtil.scala:82]: How about `actualJsonStr` and `expectedJsonStr`?

File: testkit/src/mill/testkit/BspServerUtil.scala
Comment (lihaoyi) [inline testkit/src/mill/testkit/BspServerUtil.scala:110]: `jsonFile` needs a better name. `serverMetadataFile`?

**Comment by lihaoyi** [comment]
Using the term `fixture` seems a weird here. When I think of "test fixtures", what comes to mind is the helper methods and utils code that helps set up a tear down the test, whereas in this case you are using it to refer to the actual expected test data. Can we find a different name for these things, perhaps calling them `expectedResponses` or something?

File: bsp/worker/src/mill/bsp/worker/SyntheticRootBspBuildTargetData.scala
Comment (lihaoyi) [inline bsp/worker/src/mill/bsp/worker/SyntheticRootBspBuildTargetData.scala:17]: We're removing the UUID here. Do we know that it is not necessary?

File: testkit/src/mill/testkit/BspServerUtil.scala
Comment (lihaoyi) [inline testkit/src/mill/testkit/BspServerUtil.scala:159]: 2s seems a bit of an odd timeout; it's short enough that it could conceivably cause flakiness in test under CPU/memory/GC pressure, but long enough it won't during development when tests are typically run alone. If we just want to make sure it doesn't hang forever, maybe a longer timeout like 10s or 30s? That should be long enough it will only trigger if something is really-truly stuck.

File: bsp/worker/src/mill/bsp/worker/State.scala
Comment (alexarchambault) [reply bsp/worker/src/mill/bsp/worker/State.scala:10]: Yes. Without this, the order of build targets in responses to `workspaceBuildTargets` would sometimes change over time.

File: build.mill
Comment (alexarchambault) [reply build.mill:420]: Right, that's a debug leftover

File: bsp/worker/src/mill/bsp/worker/MillBuildServer.scala
Comment (alexarchambault) [reply bsp/worker/src/mill/bsp/worker/MillBuildServer.scala:781]: Sure

File: testkit/src/mill/testkit/BspServerUtil.scala
Comment (alexarchambault) [reply testkit/src/mill/testkit/BspServerUtil.scala:36]: Yes. All bsp4j classes get serialized / deserialized well out-of-the-box with Gson.

Comment (alexarchambault) [reply testkit/src/mill/testkit/BspServerUtil.scala:36]: And bsp4j itself pulls gson in the class path.

File: testkit/src/mill/testkit/BspServerUtil.scala
Comment (alexarchambault) [reply testkit/src/mill/testkit/BspServerUtil.scala:107]: It can. Having it could be useful for more advanced tests, but the parameter can always be added back at that time.

File: testkit/src/mill/testkit/BspServerUtil.scala
Comment (alexarchambault) [reply testkit/src/mill/testkit/BspServerUtil.scala:168]: It's not only paths, it's paths and Java version. What about `normalizeLocalValuesForTesting`?

File: testkit/src/mill/testkit/BspServerUtil.scala
Comment (alexarchambault) [reply testkit/src/mill/testkit/BspServerUtil.scala:20]: Sure

File: testkit/src/mill/testkit/BspServerUtil.scala
Comment (alexarchambault) [reply testkit/src/mill/testkit/BspServerUtil.scala:34]: Sure (although bin compat might be harder to preserve with the `object`)

File: testkit/src/mill/testkit/BspServerUtil.scala
Comment (alexarchambault) [reply testkit/src/mill/testkit/BspServerUtil.scala:51]: Or `expectedValueOpt`, so that we avoid the term "fixture"?

**Comment by alexarchambault** [comment]
> Using the term `fixture` seems a weird here. When I think of "test fixtures", what comes to mind is the helper methods and utils code that helps set up a tear down the test, whereas in this case you are using it to refer to the actual expected test data. Can we find a different name for these things, perhaps calling them `expectedResponses` or something?

In [snapshot4s](https://github.com/siriusxm/snapshot4s), they call those… "snapshots"

File: bsp/worker/src/mill/bsp/worker/SyntheticRootBspBuildTargetData.scala
Comment (alexarchambault) [reply bsp/worker/src/mill/bsp/worker/SyntheticRootBspBuildTargetData.scala:17]: My guess is that this avoid conflicts with an actual module users created. But a collision seems quite unlikely.

File: build.mill
Comment (alexarchambault) [reply build.mill]: `prependShellScript` was a task only used the `launcher` task. I needed the output of `prependShellScript` to hard-code (in a `milll.main.cli` Java property) the path of the launcher generated by `launcher`. Somehow, I needed `launcher` to pass an argument to `prependShellScript` (an argument involving `T.dest`), so I initially inlined `prependShellScript`. The new push makes it stop being a task, and makes it a standard method accepting arguments instead.

File: testkit/src/mill/testkit/BspServerUtil.scala
Comment (alexarchambault) [reply testkit/src/mill/testkit/BspServerUtil.scala]: Renamed it to `bspMetadataFile` in the latest push

File: testkit/src/mill/testkit/BspServerTestUtil.scala
Comment (lihaoyi) [inline testkit/src/mill/testkit/BspServerTestUtil.scala:1]: `testkit` is meant for things that we publish for the rest of the world to use. Given that this stuff is for testing our own BSP implementation, it should probably just live in that module's test suite.

**Review by lihaoyi — APPROVED** [review]
Left one more comment but otherwise looks reasonable

File: integration/ide/bsp-server/src/BspServerTests.scala
Comment (lihaoyi) [inline integration/ide/bsp-server/src/BspServerTests.scala:19]: We should rename this to `snapshotTests` or something

Comment (alexarchambault) [reply integration/ide/bsp-server/src/BspServerTests.scala:19]: Or `requests`? This tests that the main BSP requests work fine. `project structure` might work too, this tests the "descriptive" BSP requests, that expose the project structure somehow.

Comment (lihaoyi) [reply integration/ide/bsp-server/src/BspServerTests.scala:19]: `requestSnapshotTests`? I think being a bit verbose here is fine, since the suite is definitely doing something unusual

File: testkit/src/mill/testkit/BspServerTestUtil.scala
Comment (alexarchambault) [reply testkit/src/mill/testkit/BspServerTestUtil.scala:1]: Sure. I was thinking this might be useful in other tests (multi-module ones maybe), but we can always move that elsewhere later if needed.

File: integration/ide/bsp-server/src/BspServerTests.scala
Comment (alexarchambault) [reply integration/ide/bsp-server/src/BspServerTests.scala]: That would be `requestSnapshots` in the former case, right? No need of the `Test` suffix

File: testkit/package.mill
Comment (lefou) [inline testkit/package.mill:10]: Why got `testkit` new dependencies? I can't see how this is needed.

File: integration/ide/bsp-server/src/BspServerTests.scala
Comment (alexarchambault) [reply integration/ide/bsp-server/src/BspServerTests.scala]: Renamed them the test to `requestSnapshots`

File: bsp/worker/src/mill/bsp/worker/MillBuildServer.scala
Comment (alexarchambault) [reply bsp/worker/src/mill/bsp/worker/MillBuildServer.scala:782]: Should be fixed in the latest push

File: testkit/package.mill
Comment (alexarchambault) [reply testkit/package.mill]: Should be fixed now

**Comment by lefou** [comment]
@alexarchambault It would really help if you could just push changes by new commits. If it's so important to you to rebase, please do it in the end, after we reviewed the pull request and approved it. We will most likely squash then anyways, so what's the point? 

And finally, if anyone really want's to `--force` a `git push`, please use `--force-with-lease` instead. Otherwise you ruin your co-workers `git` experience, and I think also GitHub review UI suffers a bit. I sometimes can't finish my started review, because of the fact the commits no longer exists. That's not fun.

**Comment by alexarchambault** [comment]
> It would really help if you could just push changes by new commits. If it's so important to you to rebase

I wasn't aware of force pushes being a hindrance to started reviews. (But wouldn't waiting for the CI to be green make that unlikely to happen?)

Will try to avoid force push in the future (by pushing more "fixup" commits I guess). Rebasing helps me make sense of what I do. (In my own projects, I don't mind force pushes that allow to have a clean git history, quite the opposite. I find it much easier to review the final result via a clean git history than reviewing "diffs-of-diffs", that things such as [this](https://github.com/com-lihaoyi/mill/compare/2def6ea4179a71d397304fec39e3bfcd63469c51..324b5f01ff6099caf038e2ca14d7e863089902c1) are, especially for non trivial PRs)

I've always been using `--force-with-lease` though.

**Comment by lihaoyi** [comment]
Once this is green we can merge it and rebase all the other downstream PRs for easier reviewing, I expect those will be easy to review/merge as well

**Comment by alexarchambault** [comment]
(If the CI is green, feel free to review, and possibly un-draft and merge)

File: build.mill
Comment (lihaoyi) [inline build.mill:806]: @alexarchambault what are these changes about? They don't seem related to BSP integration tests and their purpose doesn't seem obvious at a glance

File: integration/package.mill
Comment (lihaoyi) [inline integration/package.mill:81]: Let's call this `IdeIntegrationCrossModule` just to be a bit more verbose and follow the existing convention

File: build.mill
Comment (alexarchambault) [reply build.mill:806]: That's needed for [that step](https://github.com/com-lihaoyi/mill/blob/8b063c48d76837870af188ad48500bb739668f2b/bsp/src/mill/bsp/BSP.scala#L79-L83) of `mill.bsp.BSP/install` to work. This ensures that `mill.main.cli` has a valid value (path to the launcher), so that spawning a server from `.bsp/mill-bsp.json` works. Without this, `.bsp/mill-bsp.json` has `java.class.path` as executable, which doesn't work in the test (as that value is a `:`-separated list of JARs and directories, and not the launcher - the "pathing JARs" isn't there, only its list of JARs / directories)

File: integration/package.mill
Comment (alexarchambault) [reply integration/package.mill:81]: Sure

File: build.mill
Comment (lihaoyi) [reply build.mill:806]: If we want to change where `mill.main.cli` points to, could we override it by passing in the `$MILL_MAIN_CLI` CLI variable? That would be something we can do directly in the integration test without needing to patch the launcher script

Comment (lihaoyi) [reply build.mill:806]: That would let us keep `prependShellScript` as a `Task`, which would reduce the diff and keep things more easily inspectable

Comment (alexarchambault) [reply build.mill]: Sure, the latest push does that

File: testkit/src/mill/testkit/IntegrationTester.scala
Comment (lihaoyi) [inline testkit/src/mill/testkit/IntegrationTester.scala:98]: Rather than patching this here, can we duplicate `TEST_SCALA_2_13_VERSION` as a `forkEnv` in `integration/package.mill`? This seems like a Mill-internal concern and shouldn't need to be published for external users of `mill-testkit`

Comment (alexarchambault) [reply testkit/src/mill/testkit/IntegrationTester.scala]: Should be fixed

Comment (lihaoyi) [reply testkit/src/mill/testkit/IntegrationTester.scala]: looks good, once it's green let's land this

**Pull request merged**
